lucille-core/src/main/java/com/kmwllc/lucille/stage/ApplyJavascript.java

```diff
@@ -0,0 +1,408 @@
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.*;
+import com.kmwllc.lucille.connector.FileConnector;
+import com.kmwllc.lucille.core.Document;
+import com.kmwllc.lucille.core.Stage;
+import com.kmwllc.lucille.core.StageException;
+import com.kmwllc.lucille.core.UpdateMode;
+import com.kmwllc.lucille.core.spec.Spec;
+import com.kmwllc.lucille.core.spec.SpecBuilder;
+import com.kmwllc.lucille.util.FileContentFetcher;
+import com.typesafe.config.Config;
+import org.graalvm.polyglot.*;
+import org.graalvm.polyglot.proxy.ProxyArray;
+import org.graalvm.polyglot.proxy.ProxyObject;
+
+import java.io.Reader;
+import java.util.*;
+
+public class ApplyJavascript extends Stage {
+
+  public static final Spec SPEC = SpecBuilder.stage()
+      .optionalString("script_path")
+      .optionalString("script")
+      .optionalParent(FileConnector.S3_PARENT_SPEC, FileConnector.GCP_PARENT_SPEC, FileConnector.AZURE_PARENT_SPEC)
+      .build();
+
+  private final String scriptPath;
+  private final String inlineScript;
+  private final FileContentFetcher fileFetcher;
+  private Source source;
+  private Context context;
+
+  public ApplyJavascript(Config config) {
+    super(config);
+    this.scriptPath = config.hasPath("script_path") ? config.getString("script_path") : null;
+    this.inlineScript = config.hasPath("script") ? config.getString("script") : null;
+    this.fileFetcher = new FileContentFetcher(config);
+  }
+
+  @Override
+  public void start() throws StageException {
+    if ((scriptPath == null && inlineScript == null) || (scriptPath != null && inlineScript != null)) {
+      throw new StageException("Can only specify either script path or script.");
+    }
+
+    if (inlineScript != null) {
+      try {
+        this.source = Source.newBuilder("js", inlineScript, "<inline>").build();
+      } catch (Exception e) {
+        throw new StageException("Failed to build inline JavaScript source.", e);
+      }
+    } else {
+      try {
+        fileFetcher.startup();
+      } catch (Exception e) {
+        throw new StageException("Error initializing FileContentFetcher.", e);
+      }
+      try (Reader reader = fileFetcher.getReader(scriptPath)) {
+        this.source = Source.newBuilder("js", reader, scriptPath).build();
+      } catch (Exception e) {
+        throw new StageException("Failed to read JavaScript from '" + scriptPath + "'.", e);
+      }
+    }
+
+    try {
+      Engine engine = Engine.newBuilder()
+          .option("engine.WarnInterpreterOnly", "false")
+          .option("log.level", "OFF")
+          .build();
+
+      this.context = Context.newBuilder("js")
+          .engine(engine)
+          .allowAllAccess(true)
+          .build();
+    } catch (Exception e) {
+      throw new StageException("Failed to initialize JavaScript context.", e);
+    }
+  }
+
+  @Override
+  public void stop() throws StageException {
+    if (context != null) {
+      context.close();
```

Comment:
the docs don't say that closing the context will close the engine it's using
and Engine has it's own close method
so I think we have to close the engine after closing the context